core/tee/tee_fs_key_manager.c

```diff
@@ -0,0 +1,417 @@
+/*
+ * Copyright (c) 2015, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+/* Acronyms:
+ *
+ * FEK - File Encryption Key
+ * SST - Secure Storage
+ * SSK - Secure Storage Key
+ * IV  - Initial vector
+ * HUK - Hardware Unique Key
+ * RNG - Random Number Generator
+ *
+ * */
+
+#include <initcall.h>
+#include <stdlib.h>
+#include <string.h>
+#include <kernel/tee_common_otp.h>
+#include <kernel/tee_common_unpg.h>
+#include <tee/tee_cryp_utl.h>
+#include <tee/tee_cryp_provider.h>
+#include <tee/tee_fs_key_manager.h>
+#include <compiler.h>
+#include <trace.h>
+
+struct tee_fs_ssk {
+	int is_init;
+	uint8_t key[TEE_FS_KM_SSK_SIZE];
+};
+
+struct aad {
+	const uint8_t *encrypted_key;
+	const uint8_t *iv;
+};
+
+struct km_header {
+	struct aad aad;
+	uint8_t *tag;
+};
+
+
+struct common_header {
+	uint8_t iv[TEE_FS_KM_IV_LEN];
+	uint8_t tag[TEE_FS_KM_MAX_TAG_LEN];
+};
+
+struct meta_header {
+	uint8_t encrypted_key[TEE_FS_KM_FEK_SIZE];
+	struct common_header common;
+};
+
+struct block_header {
+	struct common_header common;
+};
+
+static struct tee_fs_ssk tee_fs_ssk;
+static uint8_t string_for_ssk_gen[] = "ONLY_FOR_tee_fs_ssk";
+
+
+static TEE_Result fek_crypt(TEE_OperationMode mode,
+		uint8_t *key, int size)
+{
+	TEE_Result res;
+	uint8_t *ctx = NULL;
+	size_t ctx_size;
+	uint8_t dst_key[TEE_FS_KM_FEK_SIZE];
+
+	if (!key)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (size != TEE_FS_KM_FEK_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (tee_fs_ssk.is_init == 0)
+		return TEE_ERROR_GENERIC;
+
+	res = crypto_ops.cipher.get_ctx_size(TEE_FS_KM_ENC_FFK_ALG, &ctx_size);
```

Comment:
I've been wondering if this should be `..._FFK_...` or is it a typo and should be `...FEK...`?